src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -34,59 +94,130 @@ class SortedSegmentSearch {
-        case PredicateCondition::GreaterThan: _begin = _get_last_bound(); return;
-        case PredicateCondition::LessThanEquals: _end = _get_last_bound(); return;
-        case PredicateCondition::LessThan: _end = _get_first_bound(); return;
+        case PredicateCondition::GreaterThanEquals: _begin = _get_first_bound(_first_search_value, _begin, _end); return; // NOLINT
+        case PredicateCondition::GreaterThan: _begin = _get_last_bound(_first_search_value, _begin, _end); return;
+        case PredicateCondition::LessThanEquals: _end = _get_last_bound(_first_search_value, _begin, _end); return;
+        case PredicateCondition::LessThan: _end = _get_first_bound(_first_search_value, _begin, _end); return;
         default: Fail("Unsupported predicate condition encountered");
       }
     } else {
       switch (_predicate_condition) {
-        case PredicateCondition::LessThanEquals: _begin = _get_first_bound(); return;
-        case PredicateCondition::LessThan: _begin = _get_last_bound(); return;
-        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(); return;
-        case PredicateCondition::GreaterThan: _end = _get_first_bound(); return;
+        case PredicateCondition::LessThanEquals: _begin = _get_first_bound(_first_search_value, _begin, _end); return;
+        case PredicateCondition::LessThan: _begin = _get_last_bound(_first_search_value, _begin, _end); return;
+        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(_first_search_value, _begin, _end); return;
+        case PredicateCondition::GreaterThan: _end = _get_first_bound(_first_search_value, _begin, _end); return;
         default: Fail("Unsupported predicate condition encountered");
       }
     }
     // clang-format on
   }
 
+  // This function sets the offset(s) that delimit the result set based on the predicate condition and the sort order
+  void _set_begin_and_end_positions_for_between_scan() {
+    DebugAssert(_second_search_value, "Second Search Value must be set for between scan");
+    if (_begin == _end) return;
+
+    auto first_value = _begin->value();
+    auto last_value = (_end - 1)->value();
+    if (!_is_ascending) std::swap(first_value, last_value);
+
+    // early out everything matches
+    if (first_value > _first_search_value && last_value < *_second_search_value) return;
+
+    // early out nothing matches
+    if (first_value > *_second_search_value || last_value < _first_search_value) {
+      _begin = _end;
+      return;
+    }
+
+    // This implementation uses behaviour which resembles std::equal_range's
+    // behaviour since it also calculates a range containing all elements between
+    // the first and second search value. However, std::equal_range returns all
+    // elements in range [first, last), which describes only the
+    // PredicateCondition::BetweenInclusive case. For the other PredicateConditions,
+    // different borders are required.
```

Comment:
Is this really correct? If I am not mistaken, even BetweenInclusive cannot be handled by equal_range.

How about this: "However, std::equal_range compares the values using the equals operator and a single comparison value. We need a range on the comparison side, too."

Reply:
Good catch! I added a note that we will be comparing to two different search values, if two are given.